src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -773,90 +762,52 @@ DeferredErrorIfUnsupportedRecurringTuplesJoin(
 	List *joinRestrictionList =
 		plannerRestrictionContext->joinRestrictionContext->joinRestrictionList;
 	ListCell *joinRestrictionCell = NULL;
-	RecurringTuplesType recurType = RECURRING_TUPLES_INVALID;
 	foreach(joinRestrictionCell, joinRestrictionList)
 	{
 		JoinRestriction *joinRestriction = (JoinRestriction *) lfirst(
 			joinRestrictionCell);
 		JoinType joinType = joinRestriction->joinType;
 		PlannerInfo *plannerInfo = joinRestriction->plannerInfo;
-		Relids innerrelRelids = joinRestriction->innerrelRelids;
-		Relids outerrelRelids = joinRestriction->outerrelRelids;
 
-		if (joinType == JOIN_SEMI || joinType == JOIN_ANTI || joinType == JOIN_LEFT)
-		{
-			/*
-			 * If there are only recurring tuples on the inner side of a join then
-			 * we can push it down, regardless of whether the outer side is
-			 * recurring or not. Otherwise, we check the outer side for recurring
-			 * tuples.
-			 */
-			if (RelationInfoContainsOnlyRecurringTuples(plannerInfo, innerrelRelids))
-			{
-				continue;
-			}
-
-
-			/*
-			 * If the outer side of the join doesn't have any distributed tables
-			 * (e.g., contains only recurring tuples), Citus should not pushdown
-			 * the query. The reason is that recurring tuples on every shard would
-			 * be added to the result, which is wrong.
-			 */
-			if (RelationInfoContainsOnlyRecurringTuples(plannerInfo, outerrelRelids))
-			{
-				/*
-				 * Find the first (or only) recurring RTE to give a meaningful
-				 * error to the user.
-				 */
-				recurType = FetchFirstRecurType(plannerInfo, outerrelRelids);
+		/*
+		 * This loop aims to determine whether this join is between a recurring
+		 * rel and a non-recurring rel, and if so, whether it can yield an incorrect
+		 * result set due to recurring tuples.
+		 *
+		 * For outer joins, this is not a problem when the inner side is recurring
```

Comment:
I feel like this would mean bringing the outer join checks that we dropped in this PR already back--by guarding with USE_ASSERT_CHECKING, but I'm not sure if this would worth.